diff --git include/openssl/tls1.h include/openssl/tls1.h
index 7cb6a0e..4165458 100644
--- include/openssl/tls1.h
+++ include/openssl/tls1.h
@@ -240,9 +240,6 @@ extern "C" {
 /* ExtensionType value from RFC 5620. */
 #define TLSEXT_TYPE_heartbeat	15
 
-/* ExtensionType value from RFC 7301. */
-#define TLSEXT_TYPE_application_layer_protocol_negotiation 16
-
 /* ExtensionType value for TLS padding extension.
  * (TEMPORARY - registered 2014-03-12, expires 2015-03-12)
  * http://tools.ietf.org/html/draft-agl-tls-padding-03
