diff --git src/plugins/auth-pam/auth-pam.c src/plugins/auth-pam/auth-pam.c
index bd71792..34635a7 100644
--- src/plugins/auth-pam/auth-pam.c
+++ src/plugins/auth-pam/auth-pam.c
@@ -663,6 +663,12 @@ pam_auth (const char *service, const struct user_pass *up)
   conv.conv = my_conv;
   conv.appdata_ptr = (void *)up;
   status = pam_start (service, name_value_list_provided ? NULL : up->username, &conv, &pamh);
+
+  if (status == PAM_SUCCESS)
+    {
+      status = pam_set_item(pamh, PAM_RHOST, "127.0.0.1");
+    }
+
   if (status == PAM_SUCCESS)
     {
       /* Call PAM to verify username/password */
